packages/ozone/src/daemon/event-pusher.ts

```diff
@@ -0,0 +1,295 @@
+import { InputSchema } from '../lexicon/types/com/atproto/admin/updateSubjectStatus'
+import { Selectable } from 'kysely'
+
+type EventSubject = InputSchema['subject']
+
+type PollState = {
+  timer?: NodeJS.Timer
+  promise: Promise<void>
+  tries: number
+}
+
+type AuthHeaders = {
+  headers: {
+    authorization: string
+  }
+}
+
+type Service = {
+  agent: AtpAgent
+  did: string
+}
+
+export class EventPusher {
+  destroyed = false
+
+  repoPollState: PollState = {
+    promise: Promise.resolve(),
+    tries: 0,
+  }
+  recordPollState: PollState = {
+    promise: Promise.resolve(),
+    tries: 0,
+  }
+  blobPollState: PollState = {
+    promise: Promise.resolve(),
+    tries: 0,
+  }
+
+  appview: Service | undefined
+  pds: Service | undefined
+
+  constructor(
+    public db: Database,
+    public createAuthHeaders: (aud: string) => Promise<AuthHeaders>,
+    services: {
+      appview?: {
+        url: string
+        did: string
+      }
+      pds?: {
+        url: string
+        did: string
+      }
+    },
+  ) {
+    if (services.appview) {
+      this.appview = {
+        agent: new AtpAgent({ service: services.appview.url }),
+        did: services.appview.did,
+      }
+    }
+    if (services.pds) {
+      this.pds = {
+        agent: new AtpAgent({ service: services.pds.url }),
+        did: services.pds.did,
+      }
+    }
+  }
+
+  start() {
+    this.poll(this.repoPollState, () => this.pushRepoEvents())
+    this.poll(this.recordPollState, () => this.pushRecordEvents())
+    this.poll(this.blobPollState, () => this.pushBlobEvents())
+  }
+
+  poll(state: PollState, fn: () => Promise<boolean>) {
+    if (this.destroyed) return
+    state.promise = fn()
+      .then((hadEvts: boolean) => {
+        if (hadEvts) {
+          state.tries = 0
+        } else {
+          state.tries++
+        }
+      })
+      .catch((err) => {
+        dbLogger.error({ err }, 'event push failed')
+        state.tries++
+      })
+      .finally(() => {
+        state.timer = setTimeout(
+          () => this.poll(state, fn),
+          exponentialBackoff(state.tries),
+        )
+      })
+  }
+
+  async processAll() {
+    await Promise.all([
+      this.pushRepoEvents(),
+      this.pushRecordEvents(),
+      this.pushBlobEvents(),
+    ])
+  }
+
+  async destroy() {
+    this.destroyed = true
+    const destroyState = (state: PollState) => {
+      if (state.timer) {
+        clearTimeout(state.timer)
+      }
+      return state.promise
+    }
+    await Promise.all([
+      destroyState(this.repoPollState),
+      destroyState(this.recordPollState),
+      destroyState(this.blobPollState),
+    ])
+  }
+
+  async pushRepoEvents() {
+    return await this.db.transaction(async (dbTxn) => {
+      const toPush = await dbTxn.db
+        .selectFrom('repo_push_event')
+        .selectAll()
+        .forUpdate()
+        .skipLocked()
+        .where('confirmedAt', 'is', null)
+        .where('attempts', '<', 10)
```

Comment:
The event pusher will retry each push 10 times and then give up

We may want to add smarter logic here to handle the case of downtime on the service it's pushing too

Reply:
I suppose as an escape hatch we could always reset the attempts on events that failed during an outage.

Reply:
yeah that was basically my plan in the interim 👍